feat: `add array/base/linspace2d`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   add `array/base/linspace2d` including relevant tests, benchmarks and documentation.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/19

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add array/base/linspace2d

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented:
@kgryte I pushed these changes with `--no-verify` flag. Reason being the recent migrate change https://github.com/stdlib-js/stdlib/pull/5458.

I was getting an error related to `namespace` I believe it needs to be updated?

headlessNode committed:
fix: idx and docs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented on lib/node_modules/@stdlib/array/base/linspace2d/README.md:
```diff
@@ -0,0 +1,128 @@
+
+## Notes
+
+-   The function assumes that the product of number of rows and columns is greater than or equal to `2`.
```
```suggestion
-   The function assumes that the product of the number of rows and columns is greater than or equal to `2`.
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/array/base/linspace2d/docs/repl.txt:
```diff
@@ -0,0 +1,32 @@
+        Array shape.
+
+    colexicographic: boolean
+        specifies whether the array values should be in colexicographic order
```
```suggestion
        Specifies whether the array values should be in colexicographic order
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/array/base/linspace2d/docs/types/index.d.ts:
```diff
@@ -0,0 +1,51 @@
+* @param stop - last array value
+* @param shape - array shape
+* @param colexicographic - specifies whether the array values should be in colexicographic order
+* @returns linearly spaced numeric array
```
```suggestion
* @returns linearly spaced two-dimensional nested numeric array
```

headlessNode commented on lib/node_modules/@stdlib/array/base/linspace2d/lib/main.js:
```diff
@@ -0,0 +1,88 @@
+* @param {number} stop - last array value
+* @param {NonNegativeIntegerArray} shape - array shape
+* @param {boolean} colexicographic - specifies whether the array values should be in colexicographic order
+* @returns {Array} linearly spaced numeric array
```
```suggestion
* @returns {Array} linearly spaced two-dimensional nested numeric array
```

headlessNode commented on lib/node_modules/@stdlib/array/base/linspace2d/benchmark/benchmark.size.js:
```diff
@@ -0,0 +1,95 @@
+* @private
+*/
+function main() {
+	var N;
+	var min;
+	var max;
+	var f;
+	var i;
```
```suggestion
	var min;
	var max;
	var N;
	var f;
	var i;
```

headlessNode committed:
fix: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
refactor: remove the extra loop

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode committed:
refactor: optimize the function

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented:
/stdlib merge

kgryte commented on lib/node_modules/@stdlib/array/base/linspace2d/README.md:
```diff
@@ -0,0 +1,128 @@
+    var arr = linspace( 0, 10, [ 3, 3 ], false );
+
+    // Round each value to the nearest hundredth:
+    var i;
```
Use [`map2d`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/array/base/map2d) here in order to avoid nested loops. This will also allow you to use a `// returns` annotation, rather than `console.log` here.

kgryte commented on lib/node_modules/@stdlib/array/base/linspace2d/README.md:
````diff
@@ -0,0 +1,128 @@
+x = linspace2d( 0, 100, [ 2, 3 ], true );
+// returns [ [ 0, 40, 80 ], [ 20, 60, 100 ] ]
+```
+
````
We should document the parameters here.

kgryte commented on lib/node_modules/@stdlib/array/base/linspace2d/docs/repl.txt:
```diff
@@ -0,0 +1,32 @@
+        Array shape.
+
+    colexicographic: boolean
+        Specifies whether the array values should be in colexicographic order
```
```suggestion
        Specifies whether generated array values should be stored in colexicographic order.
```

Applies here and elsewhere.

kgryte commented on lib/node_modules/@stdlib/array/base/linspace2d/docs/types/index.d.ts:
```diff
@@ -0,0 +1,51 @@
+* @returns linearly spaced two-dimensional nested numeric array
+*
+* @example
+* var linspace2d = require( '@stdlib/array/base/linspace2d' );
+*
```
```suggestion
```

I am not sure where this practice of requiring the package in TS declarations started, but this needs to be fixed.

kgryte commented on lib/node_modules/@stdlib/array/base/linspace2d/lib/index.js:
```diff
@@ -0,0 +1,43 @@
+*
+* var x = linspace2d( 0, 100, [ 2, 3 ], false );
+* // returns [ [ 0, 20, 40 ], [ 60, 80, 100 ] ]
+
```
```suggestion
*
```

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into linspace2d-array-base

kgryte commented on lib/node_modules/@stdlib/array/base/linspace2d/lib/main.js:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	d = ( stop - start ) / ( n - 1 );
+	out = zeros2d( shape );
```
You've effectively implemented this as a two-pass algorithm; however, you can implement this in a single pass. The secret is borrowing ideas from the ndarray kernels. Namely, based on the value of the colexicographic argument, you set two increment variables: one for the inner loop and one for the outer loop. The inner loop controls how values are incremented along a row. The outer loop variable controls how to increment to the first value along that row.

Using this approach, you can also remove the inner `colexicographic` branch.

kgryte commented on lib/node_modules/@stdlib/array/base/linspace2d/test/test.js:
```diff
@@ -0,0 +1,105 @@
+
+	actual = linspace2d( start, stop, sh, false );
+
+	// Verify that the output is a nested array:
```
Did you add these comments or did an LLM?

headlessNode replied:
Took a quick look at the recording, I copied the test file of `linspace` and modified the comments myself. The issue was no AI.

headlessNode committed:
docs: refactor readme

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
refactor: make single pass

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
@kgryte Changes applied. Please check.

kgryte commented on lib/node_modules/@stdlib/array/base/linspace2d/README.md:
```diff
@@ -0,0 +1,135 @@
+
+The function accepts the following arguments:
+
+-   **start**: First array value.
+-   **stop**: Last array value.
+-   **shape**: Array shape.
+-   **colexicographic**: Specifies whether generated array values should be stored in colexicographic order.
```
```suggestion
-   **start**: first array value.
-   **stop**: last array value.
-   **shape**: array shape.
-   **colexicographic**: specifies whether generated array values should be stored in colexicographic order.
```

No capitalizing descriptions following a colon.

kgryte commented on lib/node_modules/@stdlib/array/base/linspace2d/README.md:
```diff
@@ -0,0 +1,135 @@
+
+-   The function assumes that the product of the number of rows and columns is greater than or equal to `2`.
+
+-   The output `array` is guaranteed to include the `start` and `stop` values. Beware, however, that values between `start` and `stop` are subject to floating-point rounding errors. Hence,
```
```suggestion
-   The output array is guaranteed to include the `start` and `stop` values. Beware, however, that values between `start` and `stop` are subject to floating-point rounding errors. Hence,
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
refactor: create the output array in a single pass

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: demonstrate floating-point rounding error

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/array/base/linspace2d/README.md:
```diff
@@ -0,0 +1,134 @@
+
+## Notes
+
+-   The function assumes that the product of the number of rows and columns is greater than or equal to `2`.
+
```
```suggestion
```

The function doesn't actually make any assumptions regarding the array shape, apart from being an array-like object containing two elements.

kgryte committed:
docs: remove note

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM after clean-up. @headlessNode While you refactored the loop to use increments, you were still creating the output array in two passes. I've now updated this to perform a single pass, as I originally suggested.
